lib/src/geometry/polygon.dart

```diff
@@ -0,0 +1,140 @@
+import 'dart:math';
+import 'dart:ui';
+
+import '../../components.dart';
+import '../extensions/rect.dart';
+import '../extensions/vector2.dart';
+import 'shape.dart';
+
+/// The Polygon is defined from the center of the component and with
+/// percentages of the size of the component.
+/// Example: [[1.0, 0.0], [0.0, 1.0], [-1.0, 0.0], [0.0, -1.0]]
+/// This will form a diamond shape within the bounding size box.
+/// NOTE: Always define your shape is a clockwise fashion
+class Polygon extends Shape {
+  final List<Vector2> definition;
+
+  Polygon(
+    this.definition, {
+    Vector2 position,
+    Vector2 size,
+    double angle,
+  }) : super(position: position, size: size, angle: angle ?? 0);
+
+  /// With this helper method you can create your [Polygon] from absolute
+  /// positions instead of percentages. This helper will also calculate the size
+  /// and center of the Polygon.
+  factory Polygon.fromPositions(
+    List<Vector2> positions, {
+    double angle = 0,
+  }) {
+    final center = positions.fold<Vector2>(
+          Vector2.zero(),
+          (sum, v) => sum + v,
+        ) /
+        positions.length.toDouble();
+    final bottomRight = positions.fold<Vector2>(
+      Vector2.zero(),
+      (bottomRight, v) {
+        return Vector2(
+          max(bottomRight.x, v.x),
+          max(bottomRight.y, v.y),
+        );
+      },
+    );
+    final halfSize = bottomRight - center;
+    final definition =
+        positions.map<Vector2>((v) => (v - center)..divide(halfSize)).toList();
+    return Polygon(
+      definition,
+      position: center,
+      size: halfSize * 2,
+      angle: angle,
+    );
+  }
+
+  final _cachedScaledShape = ShapeCache<Iterable<Vector2>>();
+
+  /// Gives back the shape vectors multiplied by the size
+  Iterable<Vector2> get scaled {
+    if (!_cachedScaledShape.isCacheValid(<dynamic>[size])) {
+      _cachedScaledShape.updateCache(
+          () => definition?.map((p) => p.clone()..multiply(size / 2)),
+          <dynamic>[size.clone()]);
```

Comment:
This dynamic casting seems off place, I wonder why it is needed

Reply:
I wish, but Dart needs it for dynamic lists.

Reply:
cant you just delcare it as
```
T updateCache<F>(T update(), List<F> validCacheValues)
```

Reply:
No, because the list is of mixed types.

Reply:
T can still be dynamic if you provide a list of mixed types. unless you add to the list after it's inserted?

Reply:
Added a comment about it on the Cache class